### PR TITLE
[ConnectionPool] Add support for HTTPS request pooling

### PR DESCRIPTION
**Pull request by JerryKwan**
Here is how to try `HTTPS` connection pooling.  Currently only available when `TLS Interception` is enabled and within a single worker *(no multiprocessing support yet)*

You can use the following commands to do some tests:

# Generate required certificates
`make ca-certificates`

# Start proxy with TLS Interception enabled & with connection pool
```
python3.7 -m proxy \
--ca-key-file ca-key.pem \
--ca-cert-file ca-cert.pem \
--ca-signing-key-file ca-signing-key.pem \
--num-workers 1 \
--log-level DEBUG \
--enable-conn-pool \
--log-format "%(asctime)s - pid:%(process)d [%(levelname)-.1s] %(pathname)s.%(funcName)s:%(lineno)d - %(message)s"
```

# Client
`curl -v -x localhost:8899 --cacert ca-cert.pem https://httpbin.org/get`

### CODE REVIEW

File: proxy/core/connection/connection.py [hunk 1]
```diff
@@ -52,6 +52,7 @@ def connection(self) -> Union[ssl.SSLSocket, socket.socket]:
     def send(self, data: bytes) -> int:
         """Users must handle BrokenPipeError exceptions"""
         # logger.info(data)
+        logger.info('send to %s, data = |%s|', self.tag, data)
```
Comment (abhinavsingh) [inline proxy/core/connection/connection.py:55]: remove or comment out?

File: proxy/core/connection/connection.py [hunk 2]
```diff
@@ -66,6 +67,7 @@ def recv(
             (len(data), self.tag),
         )
         # logger.info(data)
+        logger.info('recv from %s, data = |%s|', self.tag, data)
```
Comment (abhinavsingh) [inline proxy/core/connection/connection.py:70]: same here

File: proxy/core/connection/pool.py [hunk 3]
```diff
@@ -64,6 +60,7 @@ class ConnectionPool:
     def __init__(self) -> None:
         # Pools of connection per upstream server
         self.pools: Dict[Tuple[str, int], Set[TcpServerConnection]] = {}
+        self.handshake_pools: Dict[Tuple[str, int], Dict[int, bytes]] = {}
```
Comment (abhinavsingh) [inline proxy/core/connection/pool.py:63]: A doc-string about this data-structure will be nice.  May be I can send out contributions related to docs directly into your PR.

File: proxy/core/connection/server.py [hunk 4]
```diff
@@ -27,6 +27,8 @@ def __init__(self, host: str, port: int) -> None:
         self._conn: Optional[Union[ssl.SSLSocket, socket.socket]] = None
         self.addr: Tuple[str, int] = (host, int(port))
         self.closed = True
+        self.handshake: Dict(int, bytes) = {}
```
Comment (abhinavsingh) [inline proxy/core/connection/server.py:30]: Don't think this is a valid type hint, I think only `[ ]` brackets will work.

**Review by abhinavsingh — COMMENTED** [review]
Thank you for this PR and work on https connection pool.  This not only help with within TLS interception mode but also at every place where `proxy.py` or plugins needs to establish a HTTPS connection upstream.

I'll run it locally and also try to dig into multiprocessing aspects.  But we should be able to merge this in as-in (with green CI :)) and need not wait for multiprocessing support to land within this PR.

I'll update back with my findings.  Meanwhile try to address these lint warnings.  I know they might seem annoying but are necessary and sometimes might even help you find bugs in the code.  PTAL at https://github.com/abhinavsingh/proxy.py/runs/4504222230?check_suite_focus=true

File: proxy/core/connection/connection.py (hunk 1 above)
Comment (JerryKwan) [reply proxy/core/connection/connection.py:55]: This log statement can be removed directly

File: proxy/core/connection/connection.py (hunk 2 above)
Comment (JerryKwan) [reply proxy/core/connection/connection.py:70]: Can remove too

File: proxy/core/connection/pool.py (hunk 3 above)
Comment (JerryKwan) [reply proxy/core/connection/pool.py:63]: We can remove this statement too, it not needed anymore

File: proxy/core/connection/server.py (hunk 4 above)
Comment (JerryKwan) [reply proxy/core/connection/server.py:30]: this two statement are not used any more too

self.handshake: Dict(int, bytes) = {}
self.session_id: Optional[bytes] = None

**Comment by abhinavsingh** [comment]
# Test data file missing or not found

Screen Shot 2021-12-17 at 6 07 42 PM

# Lint errors

Screen Shot 2021-12-17 at 6 08 16 PM

Screen Shot 2021-12-17 at 6 08 43 PM

# Spellcheck errors

Screen Shot 2021-12-17 at 6 10 20 PM
Screen Shot 2021-12-17 at 6 10 28 PM

Screen Shot 2021-12-17 at 6 10 37 PM

Screen Shot 2021-12-17 at 6 10 07 PM
Screen Shot 2021-12-17 at 6 09 58 PM
Screen Shot 2021-12-17 at 6 09 40 PM

**Comment by abhinavsingh** [comment]
@JerryKwan Can you check in the `tls_server_hello.data` file too.  Thank you

File: proxy/http/parser/tls.py [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+        else:
+            payload_length, = struct.unpack('!H', raw[3:5])
+            self.protocol_version
+            if length < 5 + payload_length:
+                logger.debug('incomplete data, len(raw) = %s, len(payload) = %s', length, payload_length)
+                return False, raw
+            else:
+                # parse
+                self.content_type = raw[0]
+                self.protocol_version = raw[1:3]
```
Comment (abhinavsingh) [inline proxy/http/parser/tls.py:412]: Here `self.protocol_version: Optional[TlsProtocolVersion] = None` is a `TlsProtocolVersion` type but you are assigning a byte to it here.

File: proxy/http/parser/tls.py [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+                if self.content_type == tlsContentType.HANDSHAKE:
+                    # parse handshake
+                    self.handshake = TlsHandshake()
+                    self.handshake.parse(payload)
+                return True, raw[5 + payload_length:]
+
+    def build(self) -> bytes:
+        data = bytes()
+        data += bytes([self.content_type])
+        data += self.protocol_version
```
Comment (abhinavsingh) [inline proxy/http/parser/tls.py:424]: Here too, using `self.protocol_version` as bytes while it is defined as `TlsProtocolVersion` type.

**Review by abhinavsingh — COMMENTED** [review]
PTAL.  Also add the missing `tls_server_hello.data` file.  Thank you

**Comment by abhinavsingh** [comment]
I made several changes related to list here https://github.com/JerryKwan/proxy.py/pull/4. Please take a look and merge into your merge, which in-turn will update this PR.

You will have to additionally address the comments I left before PR will become merge ready (green CI).   Let me know if you run into any issues.  Thank you.

File: proxy/http/parser/tls.py (hunk 5 above)
Comment (JerryKwan) [reply proxy/http/parser/tls.py:512]: Sorry, TlsParser is completed yet, we can declare self.protocol_version as bytes

File: proxy/http/parser/tls.py (hunk 6 above)
Comment (JerryKwan) [reply proxy/http/parser/tls.py:525]: changed self.protocol_vertion to bytes

**Comment by abhinavsingh** [comment]
@JerryKwan Looks like tests are still breaking.  Can you take a look at them?  Most of the errors seems to be in TLS parser related.

There are also lint errors in the code.  Click on "Details" to view what's wrong with them.

Screen Shot 2021-12-19 at 9 27 42 PM

Ignore `codecov` errors for now, which should be fixed once we have fixed all other issues.

Let me know if you need any help.

**Comment by JerryKwan** [comment]
@abhinavsingh I am sorry, I did't find the broken tests, would you please send the to me? thank you

**Comment by abhinavsingh** [comment]
Every PR goes through CI workflows.  Results of these checks are in the PR itself.  Are you unable to see the "check" section for some reason?  Expanding the "check" will tell you about individual checks that passed/failed.

See screenshot below for reference.

image

**Comment by JerryKwan** [comment]
@abhinavsingh TLS parsing is no needed any more ,how about we just remove them?
BTW, This is what I can see
image
image

**Comment by abhinavsingh** [comment]
> TLS parsing is no needed any more ,how about we just remove them?

Yes, I guess we can decouple it at-least.  Currently it's all part of a single file/class and separation is not obvious.

> BTW, This is what I can see

Yes, do also click on "Details" for failed (x marked) workflow runs.  That will take you to detail page of that workflow run, showing you errors that occurred.  Example, scroll to `lib` results within the `check` view and click `details`

Screen Shot 2021-12-21 at 1 29 49 AM

**Comment by abhinavsingh** [comment]
@JerryKwan Awesome looks like all has been resolved.  Docker problems are due to permission issues.  I'll take another look later today and merge if all looks good.  Thank you!!!

File: proxy/core/connection/connection.py (hunk 1 above)
Comment (abhinavsingh) [reply proxy/core/connection/connection.py:55]: I didn't get you.  By having a `logger.info` in `send`, instance will keep printing all data sent to client and servers (even encrypted data).  So we should not have this enabled.  At-best we must have `.debug`, but even in debug mode, printing all raw data gets too verbose.

There is a reason I have a commented out line above.  I uncomment it when necessary for local debugging.

File: proxy/core/connection/connection.py (hunk 2 above)
Comment (abhinavsingh) [reply proxy/core/connection/connection.py:70]: Same here, we should not keep this line.  Comment this out.

**Comment by JerryKwan** [comment]
Agree. These lines should be commented out

File: proxy/http/proxy/server.py [hunk 7]
```diff
@@ -32,6 +32,7 @@
 from ..plugin import HttpProtocolHandlerPlugin
 from ..exception import HttpProtocolException, ProxyConnectionFailed
 from ..parser import HttpParser, httpParserStates, httpParserTypes
+# from ..parser.tls import TlsParser, TlsHandshake, tlsContentType, tlsHandshakeType
```
Comment (abhinavsingh) [inline proxy/http/proxy/server.py:35]: Can you send a separate PR for `TlsParser` utility also.  We'll still need it later when we want to support HTTPS connection pooling across all connection interfaces.

File: proxy/core/connection/pool.py (hunk 3 above)
Comment (abhinavsingh) [reply proxy/core/connection/pool.py:63]: Is this not being used?  If not let's comment or remove any unused bit.

File: proxy/core/connection/server.py [hunk 8]
```diff
@@ -27,6 +27,8 @@ def __init__(self, host: str, port: int) -> None:
         self._conn: Optional[Union[ssl.SSLSocket, socket.socket]] = None
         self.addr: Tuple[str, int] = (host, int(port))
         self.closed = True
+        self.handshake: Dict[int, bytes] = {}
+        self.session_id: Optional[bytes] = None
```
Comment (abhinavsingh) [inline proxy/core/connection/server.py:31]: I am a little confused right now.  How is connection pooling even working currently.  As in who is the provider of `handshake` and `session_id` parameters.  Are we missing any file?

File: proxy/http/proxy/server.py (hunk 7 above)
Comment (JerryKwan) [reply proxy/http/proxy/server.py:35]: Okay

File: proxy/core/connection/server.py (hunk 8 above)
Comment (JerryKwan) [reply proxy/core/connection/server.py:31]: These should be removed too

**Comment by abhinavsingh** [comment]
Closing this for now as we separately merged only the base `TlsParser` skeleton in #922 .  Next we still need to cache the SSL session to successfully implement HTTPS connection resumptions.  See discussion in #748 for more details.

**Pull request closed without merging**